apps/src/pythonlab/patches/pythonScriptUtils.ts

```diff
@@ -54,59 +61,162 @@ export function writeSource(
     .forEach(folder => {
       // Create folder if it doesn't exist.
       const newPath = `${currentPath}${folder.name}`;
-      try {
-        pyodide.FS.readdir(newPath);
-      } catch (e) {
-        // Folder doesn't exist, create it.
-        pyodide.FS.mkdir(newPath);
-      }
+      createFolderIfNotExists(newPath, pyodide);
       // Recurse to write all children of the folder (files and folders).
       writeSource(source, folder.id, newPath + '/', pyodide);
     });
 }
 
-// Remove all source files from the Pyodide file system.
-// This ensures any deleted file is not available to be imported,
-// which could cause confusion.
-export function deleteSourceFiles(
+// Iterate over the pyodide file system and update the source object with any new or updated
+// csv or txt files, or new folders. If we see any other new files, send an error message.
+// In addition, delete every file in the working directory to prepare for the next run.
+// We want a clean working directory for each run.
+// TODO: determine if we want to do the following:
+// - look for deleted folders or files.
+// - send an error message if a non-csv/txt file is updated (this is currently ignored).
+// - process hidden folders (we currently ignore them).
+// Tracked in https://codedotorg.atlassian.net/browse/CT-509
+export function getUpdatedSourceAndDeleteFiles(
+  source: MultiFileSource,
+  id: string,
   pyodide: PyodideInterface,
-  source: MultiFileSource
+  sendMessage: (message: PyodideMessage) => void
 ) {
-  Object.values(source.files).forEach(file => {
-    const filePath = getFilePath(file.id, source);
-    try {
-      pyodide.FS.unlink(filePath);
-    } catch (e) {
-      // TODO: log error better. We catch this because it should not prevent
-      // future runs.
-      console.warn(`error unlinking Pyodide file ${filePath}, ${e}`);
+  const workingDir = pyodide.FS.cwd();
+  const directoryData = pyodide.FS.lookupPath(workingDir, {}).node;
+  const directoryContents = Object.values(
+    directoryData.contents
+  ) as PyodidePathContent[];
+  const newSource = {...source};
```

Comment:
I think spread syntax effectively only goes one level deep/copies references for any nested objects or arrays. Does that matter here? (Alternative: Use lodash to `cloneDeep`.)